_posts/2017-08-28-typed-notification-ios.md

```diff
@@ -0,0 +1,209 @@
+
+This is also the most important part of our `Type-Safe Notification`. As now we will have strictly defined data in our application and all the parsing logic is separated out for easy debugging and testing.
+
+{% highlight swift %}
+extension KeyboardData {
+    init(note: Notification) {
+        guard let userInfo = note.userInfo else {
+            self.beginFrame = CGRect.zero
+            self.endFrame = CGRect.zero
+            self.curve = nil
+            self.duration = 0
+
+            return
+        }
+
+        self.beginFrame = userInfo[UIKeyboardFrameBeginUserInfoKey] as! CGRect
+        self.endFrame = userInfo[UIKeyboardFrameEndUserInfoKey] as! CGRect
+        self.curve = UIViewAnimationCurve(rawValue: (userInfo[UIKeyboardAnimationCurveUserInfoKey] as? Int) ?? 0)
+        self.duration = userInfo[UIKeyboardAnimationDurationUserInfoKey] as! TimeInterval
+    }
+}
+{% endhighlight %}
+
+`KeboardShowDescriptor` and `KeyboardHideDescriptor` will become
+
+{% highlight swift %}
+struct SystemNotification {
+    static let keyboardShowNotification = NotificationDescriptor<KeyboardData>(name: Notification.Name.UIKeyboardWillShow, convert: KeyboardData.init)
+    static let keyboardHiderNotification = NotificationDescriptor<KeyboardData>(name: Notification.Name.UIKeyboardWillHide, convert: KeyboardData.init)
+}
+{% endhighlight %}
+
+We have also encapsulated these properties in `SystemNotification` struct, so that our global namespace doesn't get polluted with all different types of notification descriptors.
+
+### UIViewController usage
+
+As we have these properties already define, using in a UIViewController will become very simple and we don't have to track the notification for deregister.
+
+{% highlight swift %}
+private var keyboardShowToken: Token?
+private var keyboardHideToken: Token?
+
+override func viewDidLoad() {
+    super.viewDidLoad()
+
+    registerForKeyboardNotifications()
+}
+
+func registerForKeyboardNotifications() {
+        keyboardShowToken = NotificationCenter.default.addObserver(descriptor: SystemNotification.keyboardShowNotification) { keyboardData in
+          //Custom logic related to view only
+        }
+
+        keyboardHideToken = NotificationCenter.default.addObserver(descriptor: SystemNotification.keyboardHideNotification) { keyboardData in
+            //Custom logic related to view only
+        }
+    }
+{% endhighlight %}
+
+This way we are also freeing up our UIViewController from other logic and only allowing it to handle views. This is also highly reusable as all the repetitive code is moved to a single place and this makes our code `DRY`.
+
+For MVVM architecture above implementation fits perfect, as in MVVM a viewController should only handle view related logic only.
```

Comment:
we can make MVVM a hyperlink here and provide a link which leads to a good article about MVVM